rust/lance/src/io/commit/conflict_resolver.rs

```diff
@@ -0,0 +1,722 @@
+            .operation
+            .modified_fragment_ids()
+            .collect::<HashSet<_>>();
+
+        let initial_fragments = fragments
+            .iter()
+            .filter(|fragment| {
+                // Check if the fragment is modified by the transaction.
+                modified_fragment_ids.contains(&fragment.id)
+            })
+            .map(|fragment| (fragment.id, (fragment.clone(), false)))
+            .collect::<HashMap<_, _>>();
+
+        Ok(Self {
+            initial_fragments,
+            transaction,
+            affected_rows,
+        })
+    }
+
+    fn retryable_conflict_err(
+        &self,
+        other_transaction: &Transaction,
+        other_version: u64,
+        location: Location,
+    ) -> crate::Error {
+        crate::Error::RetryableCommitConflict {
+            version: other_version,
+            source: format!(
+                "This {} transaction was preempted by concurrent transaction {} at version {}. Please retry.",
+                self.transaction.operation, other_transaction.operation, other_version).into(),
+            location,
+        }
+    }
+
+    /// Check whether the transaction conflicts with another transaction.
+    ///
+    /// Will return an error if the transaction is not valid. Otherwise, it will
+    /// return Ok(()).
+    pub fn check_txn(
+        &mut self,
+        other_transaction: Option<&Transaction>,
+        other_version: u64,
+    ) -> Result<()> {
+        let Some(other_transaction) = other_transaction else {
+            return Err(crate::Error::Internal {
+                message: format!(
+                    "There was a conflicting transaction at version {}, \
+                    and it was missing transaction metadata.",
+                    other_version
+                ),
+                location: location!(),
+            });
+        };
+
+        match self.transaction.conflicts_with(other_transaction) {
+            ConflictResult::Compatible => Ok(()),
+            ConflictResult::NotCompatible => {
+                Err(crate::Error::CommitConflict {
+                    version: other_version,
+                    source: format!(
+                        "This {} transaction is incompatible with concurrent transaction {} at version {}.",
+                        self.transaction.operation, other_transaction.operation, other_version).into(),
+                    location: location!(),
+                })
+            },
+            ConflictResult::Retryable => {
+                match &other_transaction.operation {
+                    Operation::Update { updated_fragments, removed_fragment_ids, .. } |
+                    Operation::Delete { updated_fragments, deleted_fragment_ids: removed_fragment_ids, .. } => {
+                        if self.affected_rows.is_none() {
+                            // We don't have any affected rows, so we can't
+                            // do the rebase anyways.
```

Comment:
A merge insert might not have any affected rows because it is a horizontal merge insert and I agree we have to fail at that point.

However, is it possible a delete might not have any affected rows because it only deletes whole fragments?  In that case it seems a conflict is not needed?

However, I see this as a very minor nit.  The odds of a tx we accidentally reject (deletes that delete whole fragments) seem pretty low and the impact (need to retry tx) is pretty minor.